session.go

```diff
@@ -224,24 +228,35 @@ func (s *Session) AcceptStream() (*Stream, error) {
 // Attempts to send a GoAway before closing the connection.
 func (s *Session) Close() error {
 	s.shutdownLock.Lock()
-	defer s.shutdownLock.Unlock()
+	{
+		if s.shutdown {
+			s.shutdownLock.Unlock()
+			return nil
+		}
 
-	if s.shutdown {
-		return nil
-	}
-	s.shutdown = true
-	if s.shutdownErr == nil {
-		s.shutdownErr = ErrSessionShutdown
+		s.shutdown = true
+
+		if s.shutdownErr == nil {
+			s.shutdownErr = ErrSessionShutdown
+		}
 	}
+	s.shutdownLock.Unlock()
+
+	// only one goroutine can reach here
+
 	close(s.shutdownCh)
+	<-s.sendDeadCh
```

Comment:
I don't see this channel getting closed anywhere and that makes me worried this receive operation could block forever, most likely in the future after someone inadvertently introduces a codepath that doesn't send over sendDeadCh. I need to look into the surrounding code more but this channel looks similar to recvDoneCh, but in the comment on line 71 (in your branch) you can see that the author there is guaranteeing the recvDoneCh will be closed and so the receive on recvDoneCh a few lines down will (eventually) proceed. It would be nice to have a comment about how sendDeadCh is to be used and to make guarantees about when and how many signals it will receive.